sys/kern/vm_map.c

```diff
@@ -517,6 +523,69 @@ vm_map_t *vm_map_clone(vm_map_t *map) {
   return new_map;
 }
 
+int insert_anon(vm_map_entry_t *ent, size_t off, vm_anon_t **anon,
+                vm_prot_t *ins_prot) {
+  if (*anon) {
+    /* It's already present */
+    *ins_prot = ent->prot;
+    return 0;
+  }
+
+  vm_anon_t *new = vm_anon_alloc();
+  if (!new)
+    return ENOMEM;
+
+  vm_anon_t *ret = vm_amap_insert_anon(ent->aref, new, off);
+  assert(ret == NULL);
+
+  *ins_prot = ent->prot;
+  *anon = new;
+  return 0;
+}
+
+int cow_page_fault(vm_map_t *map, vm_map_entry_t *ent, size_t off,
+                   vm_anon_t **anon, vm_prot_t *ins_prot, vm_prot_t access) {
+  /* We don't need to modify amap. We do the same as in non-cow case but we have
+   * to limit protection of page that will be inserted. */
+  if ((access & VM_PROT_WRITE) == 0) {
```

Comment:
If you moved this clause before L640 you could save some space by noticing that in all other cases `*ins_prot = ent->prot`. Maybe you could even remove ins_prot argument to both `insert_anon` and `cow_page_fault`?